presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java

```diff
@@ -0,0 +1,174 @@
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanRewriter;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.optimizations.HashGenerationOptimizer.getHashExpression;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Converts Single Distinct Aggregation into GroupBy
+ *
+ * Aggregate[a, b] => [a:type, b:type, agg:aggReturnType]
+ *    agg := "aggregation"(DISTINCT "c") (mask = c$distinct)
+ *        - MarkDistinct[distinct=a:type, b:type, c:type marker=c$distinct] => [c:type, a:type, b:type, c$distinct:boolean]
+ *
+ * AS
+ *
+ * Aggregate[a, b] => [a:type, b:type, agg:aggReturnType]
+ *    agg := "aggregation"("c")
+ *        - Project => [c:type, a:type, b:type]
+ *            - Aggregate[a, b, c] => [c:type, a:type, b:type]
+ *
+ * Project contains hash symbol when hash generation is on
+ */
+public class SingleDistinctOptimizer
+        extends PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return PlanRewriter.rewriteWith(new Optimizer(idAllocator, symbolAllocator), plan, null);
+    }
+
+    private static class Optimizer
+            extends PlanRewriter<Void>
+    {
+        private final SymbolAllocator symbolAllocator;
+        private final PlanNodeIdAllocator idAllocator;
+        private final String distinctSuffix = "distinct";
+
+        private Optimizer(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = checkNotNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = checkNotNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            if (isSingleCommonDistinct(node) == false) {
+                return node;
+            }
+
+            MarkDistinctNode distinctNode = getDistinctNode(node);
+            if (distinctNode == null) {
+                return node;
+            }
+
+            return buildAggregation(distinctNode, node);
+        }
+
+        // all aggregation functions in an aggregation node have a single common distinct mask symbol
+        private boolean isSingleCommonDistinct(AggregationNode node)
+        {
+            if (node.getMasks().size() == 1) {
```

Comment:
This would only handle a single aggregation. If there are multiple aggregations, such as in:

```
SELECT count(DISTINCT x), sum(DISTINCT x) FROM ...
```

the `node.getMasks()` map will return two entries pointing to the same symbol.